docs/profile-data.md

```diff
@@ -0,0 +1,236 @@
+          location, // index into stringTable, points to strings like:
+                    // JS: "Startup::XRE_Main"
+                    // C++: "0x7fff7d962da1"
+          // JS Only:
+          implementation, // TODO
+          optimizations, // JSON info about JIT optimizations.
+          line, // The line of code
+          category // int bitmask of the category
+            // 16 - js::ProfileEntry::Category::OTHER
+            // 32 - js::ProfileEntry::Category::CSS
+            // 64 - js::ProfileEntry::Category::JS
+            // 128 - js::ProfileEntry::Category::GC
+            // 256 - js::ProfileEntry::Category::CC
+            // 512 - js::ProfileEntry::Category::NETWORK
+            // 1024 - js::ProfileEntry::Category::GRAPHICS
+            // 2048 - js::ProfileEntry::Category::STORAGE
+            // 4096 - js::ProfileEntry::Category::EVENTS
+            // 9000 - other non-bitmask category
+        }),
+        ...
+      ],
+
+      // A table of stacks. Each entry contains references to the current frame and
+      // the parent frame (if one exists). This can then be used to reconstruct
+      // the call stack.
+
+      stackTable: SchemaData({
+        frame, // The current frame: index of the frame in the frameTable
+        prefix // The calling frame: Index of the frame in the frameTable
+      }),
+
+      // A list of arbitrary strings used within the profile.
+
+      stringTable: ["(root)", "0x109eba3b4", "Events::ProcessGeckoEvents", ...],
+
+    },
+    ...
+  ]
+}
+```
+
+The problem with this provided format for the use in Cleopatra is that it creates a lot of little objects. Because of this GC pauses become a real issue. So rather than have a bunch of small objects that get passed around, the data is mapped into a new form. For instance looking at markers the data goes through the following mapping.
+
+### Original data
+
+```js
+markers: {
+  schema: {name: 0, time: 1, data: 2},
+  data: [
+    [1, 37070.186708, { ... }],
+    [2, 37385.485834, { ... }],
+    [5, 37471.444024, { ... }],
+    ...
+  ]
+}
+```
+
+### Mapped data
+
+```js
+markers: {
+  name: [1, 2, 5, ...],
+  time: [37070.186708, 37385.485834, 37471.444024, ...],
+  data: [{...}, {...}, {...}, ...],
+}
+```
+
+While this format is a little more difficult to work with compared to mapping into the shape of `[{name, time, data}, {name, time, data}, ...]`, it has benefits in the speed of the Cleopatra client and the lack of GC pauses.
+
+### Processed Profile Format
+
+The threads of the processed profile use the above format and include the following keys:
+
+Profile data:
+
+ * samples
+ * markers
+
+Index-based table data:
+
+ * resourceTable
+ * frameTable
+ * funcTable
+ * stackTable
+ * stringTable
+
+Miscellaneous data:
+
+ * name
+ * libs
+ * tid
+
+The only information not directly taken from nsIProfiler is the funcTable which has the following properties. (TODO - describe what this represents is the the symbolication data?)
```

Comment:
The main problem here is that different "frame"s can be inside the same function, and we only discover that fact once we have the symbol table. But we also don't want to throw away the information about what parts of the function we're executing, because we may want to expose it in the UI at some point. So we make each frame point to the function that it's in.